### PR TITLE
Fix CUDA graph capture with FSDP always reshard

### PR DESCRIPTION
**Pull request by xmfan**
Stacked PRs:
 * __->__#4143

--- --- ---

## Summary
  TorchTitan groups norm and lm_head into a single FSDP unit. With chunked loss, the decoder forward runs norm but skips lm_head; ChunkedLossWrapper then invokes lm_head separately for each chunk.
```
fully_shard(
      [model.norm, model.lm_head],
      **fsdp_config,
      reshard_after_forward=reshard_after_forward_policy == "always",
  )
```

The existing override for `reshard_after_forward=False` is too late to catch the post forward that does the reshard, and some `all_gather_state` is left on the module, which contains some CUDA event outside of CUDA graph capture.

## Fix

  Explicitly unshard lm_head while FSDP is idle before processing loss chunks. This avoids populating the shared state while still reusing the unsharded weight across chunks. This shouldn't affect the overall perf, the AG wasn't prefetched in the existing code.

  The existing CUDA graph opt-out for this integration test is removed.

| Step | Operation | Before the fix | After the fix |
|---:|---|---|---|
| 1 | `norm` pre-forward | Implicitly unshards both `norm` and `lm_head`. FSDP retains the norm all-gather event in `all_gather_state`. | Same. |
| 2 | `norm` forward | `norm` runs, while `_skip_lm_head=True` prevents `lm_head` from running. The grouped FSDP forward is incomplete. | Same. |
| 3 | Decoder root post-forward | FSDP force-completes the partial group. Because the policy is `"always"`, it reshards both `norm` and `lm_head`. Root cleanup consumes and clears the norm all-gather event. | Same. |
| 4 | Enter `ChunkedLossWrapper` | The group is sharded and `all_gather_state` is empty. The wrapper sets reshard-after-forward and reshard-after-backward to `False`. This does not undo the reshard from step 3. | Same. |
| 5 | Prepare `lm_head` | No explicit unshard occurs. | The wrapper explicitly calls `lm_head.unshard()` while FSDP is idle. This unshards the entire `norm`/`lm_head` group, waits for the all-gather, and releases its event without populating forward-overlap state. |
| 6 | First `lm_head(h_chunk)` | The FSDP pre-forward hook must implicitly unshard the group. Because this happens in the forward state, FSDP retains the eager all-gather event in `all_gather_state`. | The group is already unsharded, so the pre-forward hook does not issue another all-gather and `all_gather_state` remains empty. |
| 7 | Remaining chunks | They reuse the unsharded parameters, so no later all-gather consumes the retained event. | They reuse the unsharded parameters and no all-gather event is retained. |
| 8 | Finish chunked loss | The wrapper restores the FSDP settings and calls `reshard()`. The parameters become sharded, but `reshard()` does not clear `all_gather_state`, so the eager event remains. | The wrapper restores the FSDP settings and calls `reshard()`. The parameters become sharded and `all_gather_state` remains empty. |
| 9 | Begin CUDA graph capture | The next FSDP all-gather attempts to wait on the event recorded during eager warmup, violating CUDA stream-capture isolation. | Capture begins with clean FSDP communication state and succeeds. |

  ## Testing

  - pytest -q tests/unit_tests/test_loss.py — 20 passed
  - 2-GPU, 10-step fsdp_reshard_always_spmd_types run successfully captured and replayed a CUDA graph
  - Deterministic eager and CUDA graph runs produced bitwise-identical loss and grad norm for all 10 steps

### CODE REVIEW

**Review by weifengpy — APPROVED** [review]
looks harmless. this is simply explicit calling unshard instead of letting lm_head(h_chunk) trigger unshard

**Review by tianyu-l — APPROVED** [review]

**Comment by tianyu-l** [comment]
am I reading it right that after the fix, the "always" policy would
- prefetch norm and lm_head
- perform norm
- reshard norm and lm_head
- enter chunk loss, and do explicitly unshard (used to be implicit unshard) of norm and lm_head

This sounds quite unoptimal -- should we just delete "always" policy? I guess its value is to provide a tradeoff between memory and comms when chunk loss is not used. We should at least ban "always" + chunk loss, cudagraph or not.

**Comment by xmfan** [comment]
If we disregard module boundaries for a second, a performant chunked loss config should do:

unshard norm + lm_head (together since it's a single fsdp unit)
norm
chunk on seq and compute loss
reshard

So I see it as a composability issue with FSDP/nn.Module wrappers.
